esp/services/esdl_svc_engine/esdl_monitor.cpp

```diff
@@ -0,0 +1,576 @@
+        lcid.trim().toLowerCase();
+        if (lcid.length() == 0)
+            return;
+        defId = lcid.str();
+        DBGLOG("Removing esdl definition %s from shared cache", defId);
+        {
+            CriticalBlock cb(m_critSect);
+            m_esdlInstanceMap.remove(defId);
+        }
+    }
+
+    virtual Linked<IEsdlDefinition> query(const char* defId) override
+    {
+        if (!defId || !*defId)
+            return nullptr;
+        StringBuffer lcid(defId);
+        lcid.trim().toLowerCase();
+        if (lcid.length() == 0)
+            return nullptr;
+        defId = lcid.str();
+        {
+            CriticalBlock cb(m_critSect);
+            Owned<CEsdlInstance>* instancepp = m_esdlInstanceMap.getValue(defId);
+            if (instancepp)
+                return (*instancepp)->m_def.get();
+        }
+        return nullptr;
+    }
+
+    //Thread periodically cleaning definition objects that are no longer being used
+    virtual int run() override
+    {
+        while (!m_stop)
+        {
+            m_waitsem.wait(300000);
+            if(m_stop)
+                break;
+            {
+                CriticalBlock cb(m_critSect);
+                StringArray ids;
+                for (auto& item : m_esdlInstanceMap)
+                {
+                    CEsdlInstance* instance = item.getValue();
+                    if (!instance || !instance->m_def.get())
+                        continue;
+                    if (!instance->m_def->isShared())
+                        ids.append(instance->m_defId.str());
+                }
+                ForEachItemIn (x, ids)
+                {
+                    const char* id = ids.item(x);
+                    DBGLOG("Definition %s is no longer being used, remove it from esdl shared cache", id);
+                    m_esdlInstanceMap.remove(id);
+                }
+            }
+        }
+
+        return 0;
+    }
+
+    virtual void stop()
+    {
+        m_stop = true;
+        m_waitsem.signal();
+    }
+private:
+    CriticalSection m_critSect;
+    MapStringTo<Owned<CEsdlInstance>> m_esdlInstanceMap;
+    Semaphore m_waitsem;
+    bool m_stop;
+};
+
+static CriticalSection gEsdlMonitorCritSection;
+
+class CEsdlMonitor : implements IEsdlMonitor, public CInterface, implements IEsdlListener
+{
+private:
+    MapStringTo<EsdlBindingImpl*> m_esdlBindingMap;
+    CriticalSection m_CritSect;
+    StringBuffer mEnvptHeader;
+    Owned<IEsdlStore> m_pCentralStore;
+    Owned<IEsdlSubscription> m_pSubscription;
+    Owned<CEsdlShare> m_esdlShare;
+    bool m_isSubscribed;
+
+public:
+    IMPLEMENT_IINTERFACE;
+
+    CEsdlMonitor() : m_isSubscribed(false)
+    {
+        constructEnvptHeader();
+        m_pCentralStore.setown(createEsdlCentralStore());
+        m_esdlShare.setown(new CEsdlShare());
+        m_esdlShare->start();
+        DBGLOG("EsdlMonitor started.");
+    }
+
+    virtual ~CEsdlMonitor()
+    {
+        if (m_pSubscription)
```

Comment:
the unsubscribe method sets m_isSubscribed=false.. it looks like this should check that before trying to unsubscribe again.

Reply:
It checks it in unsubscribe, but not in this destructor.

Reply:
yes, still looks a problem to me.